c10/util/MaybeOwned.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <c10/util/in_place.h>
+
+namespace c10 {
+
+/// A smart pointer around either a borrowed or owned T. Maintains an
+/// internal raw pointer when constructed with borrowed(), with all
+/// the attendant lifetime concerns.  Compare to Rust's
+/// std::borrow::Cow
+/// (https://doc.rust-lang.org/std/borrow/enum.Cow.html).
+template <typename T>
+class TORCH_API MaybeOwned {
+  bool isBorrowed_;
+  union {
+    const T *borrow_;
+    T own_;
+  };
+  explicit MaybeOwned(const T& t) : isBorrowed_(true), borrow_(&t) {}
```

Comment:
those are intentionally private but implemented? Do you intend MaybeOwned to be a pure temporary value used as a return type, or should people be able to keep a local copy around?

Reply:
The named static functions should be used instead of these constructors. Otherwise you might accidentally borrow a temporary.